src/Avalonia.Controls/Calendar/CalendarDatePicker.cs

```diff
@@ -538,16 +197,200 @@ protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
+                        {
+                            _calendar.DisplayMode = CalendarMode.Month;
+                        }
+
+                        if (value)
+                        {
+                            OpenDropDown();
+                        }
+                        else
+                        {
+                            _popUp.IsOpen = false;
+                            _isFlyoutOpen = _popUp.IsOpen;
+                            _isPressed = false;
+
+                            UpdatePseudoClasses();
+                            OnCalendarClosed(new RoutedEventArgs());
+                        }
+                    }
+                }
+            }
+            // SelectedDate
+            else if (change.Property == SelectedDateProperty)
+            {
+                var addedDate = change.GetNewValue<DateTime?>();
+                var removedDate = change.GetOldValue<DateTime?>();
+
+                if (SelectedDate != null)
+                {
+                    DateTime day = SelectedDate.Value;
+
+                    // When the SelectedDateProperty change is done from
+                    // OnTextPropertyChanged method, two-way binding breaks if
+                    // BeginInvoke is not used:
+                    Threading.Dispatcher.UIThread.InvokeAsync(() =>
+                    {
+                        _settingSelectedDate = true;
+                        Text = DateTimeToString(day);
+                        _settingSelectedDate = false;
+                        OnDateSelected(addedDate, removedDate);
+                    });
+
+                    // When DatePickerDisplayDateFlag is TRUE, the SelectedDate
+                    // change is coming from the Calendar UI itself, so, we
+                    // shouldn't change the DisplayDate since it will automatically
+                    // be changed by the Calendar
+                    if ((day.Month != DisplayDate.Month || day.Year != DisplayDate.Year) && (_calendar == null || !_calendar.CalendarDatePickerDisplayDateFlag))
+                    {
+                        DisplayDate = day;
+                    }
+                    if(_calendar != null)
+                        _calendar.CalendarDatePickerDisplayDateFlag = false;
+                }
+                else
+                {
+                    _settingSelectedDate = true;
+                    SetWaterMarkText();
+                    _settingSelectedDate = false;
+                    OnDateSelected(addedDate, removedDate);
+                }
+            }
+            // SelectedDateFormat
+            else if (change.Property == SelectedDateFormatProperty)
+            {
+                OnDateFormatChanged();
+            }
+            // Text
+            else if (change.Property == TextProperty)
+            {
+                var oldValue = change.GetOldValue<string>();
+                var value = change.GetNewValue<string>();
```

Comment:
minor: Now you can use
```suggestion
                var (oldValue, value) = change.GetOldAndNewValue<string>();
```

Reply:
Thanks, I never looked up all the methods and extensions that were available in this type of code (actually I think there were too many options).

Reply:
It's a new method in master branch